src/core/Akka/Pattern/BackoffOnRestartSupervisor.cs

```diff
@@ -0,0 +1,125 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="BackoffOnRestartSupervisor.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using Akka.Actor;
+using Akka.Event;
+
+namespace Akka.Pattern
+{
+    internal sealed class BackoffOnRestartSupervisor : BackoffSupervisorBase
+    {
+        private readonly TimeSpan _minBackoff;
+        private readonly TimeSpan _maxBackoff;
+        private readonly double _randomFactor;
+        private readonly OneForOneStrategy _strategy;
+        private readonly ILoggingAdapter _log = Context.GetLogger();
+
+        public BackoffOnRestartSupervisor(
+            Props childProps,
+            string childName,
+            TimeSpan minBackoff,
+            TimeSpan maxBackoff,
+            IBackoffReset reset,
+            double randomFactor,
+            OneForOneStrategy strategy) : base(childProps, childName, reset)
+        {
+            _minBackoff = minBackoff;
+            _maxBackoff = maxBackoff;
+            _randomFactor = randomFactor;
+            _strategy = strategy;
+        }
+
+        protected override SupervisorStrategy SupervisorStrategy()
+        {
+            return new OneForOneStrategy(_strategy.MaxNumberOfRetries, _strategy.WithinTimeRangeMilliseconds,
+                ex =>
+                {
+                    // TODO: StackOverflowException here
+                    //var defaultDirective = base.SupervisorStrategyInternal.AsInstanceOf<OneForOneStrategy>().Decider?.Decide(ex) ?? Directive.Escalate;
```

Comment:
My guess is big tail recursion stack here... Is there a chance that this call can be self-referential?